lib/requestwrapper.ts

```diff
@@ -192,6 +196,7 @@ export function createRequest(parameters, _callback) {
     // Remove non-valid inputs for buildRequestFileObject,
     // i.e things like {contentType: <contentType>}
     Object.keys(formData).forEach(key => {
+      // tslint:disable-next-line:no-unused-expression
```

Comment:
How about we refactor this for readability and place it into a function?

Something like this 

```javascript
function parseFormData(formData: Object): Object {
  let _formData = extend({}, formData);
  // cleanup file formData with null/undefined data
  Object.keys(_formData).forEach(key => {
    if (_formData[key].contentType && formData[key].data == null) {
      delete _formData[key];
    }
  });
  // cleanup empty objects
  Object.keys(_formData).forEach(key => {
    if (isEmptyObject(_formData[key])) {
      delete _formData[key];
    }
  });
  // convert file formData to request-style objects
  // and stringify JSON formData
  Object.keys(_formData).forEach(key => {
    if (isFileParam(_formData[key])) {
      _formData[key] = buildRequestFileObject(_formData[key]);
    } else if (typeof formData[key] === 'object') {
      _formData[key] = JSON.stringify(_formData[key]);
    }
  });
  return _formData;
}
```

Reply:
Refactoring is good, but do we even need this code in the first place? Why does typescript say this code is unused, it is a bit unclear as to its purpose

Reply:
To understand why we need this code, let's take a look at `createModel` from Language Translator.

This snippet
```javascript
const formData = {
      forced_glossary: {
        data: _params.forced_glossary,
        contentType: 'application/octet-stream'
      },
      parallel_corpus: {
        data: _params.parallel_corpus,
        contentType: 'application/octet-stream'
      },
      monolingual_corpus: {
        data: _params.monolingual_corpus,
        contentType: 'text/plain'
      }
    };
```

That first part of that code is needed to clean up things that looks like this
```javascript
{
  data: undefined,
  contentType: 'application/octet-stream'
}
```
We'll run into this situation for example if the user didn't pass the optional `forced_glossary` when they called `createModel`.

The second part that cleans up empty object literals is needed for the scenario in which we have a JSON formData parameter than end up being passed in as an empty object. The problem here is that the `requests` library will choke 👎  upon receiving those in formData, therefore we need to remove those before passing the request properties to the library. An example of this is the `parameters` parameter in Visual Recognition's `classify`.

The third part has two jobs. First is to Stringify JSON formData parameters before passing them along to `requests`, without doing so `requests` will choke 👎 . Second is responsible for constructing a `requests-style` formData object for each file parameter. This is done so that we uniformly handle how we send files to the services, despite their individual peculiarities around needing filenames or content types to complete the request. If you take a look at `buildRequestFileObject`, you'll see that it will attempt to guess the content type if it's not provided which is often important for services like VR, and it will also try to set a filename if one isn't given, which is important for services like Discovery.

Lastly, the TypeScript linter was issuing a `unused-expression` warning. I believe this is due to the linter being confused about whether some of the expressions in the code have side effects. I think if you rewrite the code with if statements, you shall no longer see those warnings. See [this](https://palantir.github.io/tslint/rules/no-unused-expression/).

Reply:
This specific code you suggested didn't work and broke tests, but this can be looked into later for improving readability